### PR TITLE
Plan reference<->local table joins locally

### PR DESCRIPTION
**Pull request by pykello**
To do this, in the planner we check if the command is a SELECT which contains both local tables and reference tables, and if so replace all reference table rtes with the shard table and plan it locally.

In the executor, we make sure that we don't do these joins in transaction blocks because we might do the changes to the reference table over MultiConnections which won't be visible to the locally planned queries. This also saves us from worrying about locking, etc. for now.

### CODE REVIEW

File: src/backend/distributed/executor/multi_executor.c [hunk 1]
```diff
@@ -631,3 +648,36 @@ AlterTableConstraintCheck(QueryDesc *queryDesc)
...
+
+	if (plannedStmt->commandType != CMD_SELECT)
+	{
+		return false;
+	}
+
+	foreach(oidCell, plannedStmt->relationOids)
+	{
+		Oid relationId = lfirst_oid(oidCell);
+		if (RelationIsAKnownShard(relationId))
```
Comment (pykello) [inline src/backend/distributed/executor/multi_executor.c]: No need to check if this is a reference table or not here. We wouldn't get here for non-reference distributed tables.

File: src/backend/distributed/planner/distributed_planner.c [hunk 2]
```diff
@@ -238,10 +256,31 @@ distributed_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 						errhint("Consider using PL/pgSQL functions instead.")));
 	}
 
+	if (localReferenceTableJoin)
+	{
+		/* See comments for IsLocalReferenceTableJoinPlan */
+		result->utilityStmt = (Node *) parse;
```
Comment (pykello) [inline src/backend/distributed/planner/distributed_planner.c]: This is pretty hacky and ugly way to do this, alternatives are (1) using CustomScan nodes (2) checking the parse tree in the executor again. 1 might be a bit more complicated to implement, 2 might slow down all of the worker queries quite a bit. This is a temporary solution and we will remove this after fixing the limitations of local execution. Let's chat about this more tomorrow.

**Comment by marcocitus** [comment]
This doesn't work correctly for `WITH ins AS (INSERT INTO ref VALUES (1,2) RETURNING *) SELECT * FROM ins, local;`

File: src/backend/distributed/executor/multi_executor.c [hunk 3]
```diff
@@ -70,6 +73,22 @@ CitusExecutorStart(QueryDesc *queryDesc, int eflags)
 {
 	PlannedStmt *plannedStmt = queryDesc->plannedstmt;
 
+	if (CitusHasBeenLoaded())
+	{
+		bool hasReferenceTableReplica = false;
+
+		PrimaryNodeForGroup(GetLocalGroupId(), &hasReferenceTableReplica);
```
Comment (marcocitus) [inline src/backend/distributed/executor/multi_executor.c]: This is clever, but really needs a comment.

**Comment by pykello** [comment]
> WITH ins AS (INSERT INTO ref VALUES (1,2) RETURNING *) SELECT * FROM ins, local;

Fixed.

File: src/include/distributed/multi_logical_planner.h [hunk 4]
```diff
@@ -213,7 +213,6 @@ extern List * TableEntryList(List *rangeTableList);
 extern List * UsedTableEntryList(Query *query);
 extern List * pull_var_clause_default(Node *node);
 extern bool OperatorImplementsEquality(Oid opno);
-extern bool FindNodeCheck(Node *node, bool (*check)(Node *));
```
Comment (pykello) [inline src/include/distributed/multi_logical_planner.h]: there were two of these in the same file.

File: src/backend/distributed/executor/multi_executor.c [hunk 5]
```diff
@@ -70,6 +73,28 @@ CitusExecutorStart(QueryDesc *queryDesc, int eflags)
...
+		 * All groups that have pg_dist_node entries, also have reference
+		 * table replicas.
+		 */
+		PrimaryNodeForGroup(GetLocalGroupId(), &hasReferenceTableReplica);
+
+		if (IsCoordinator() &&
+			hasReferenceTableReplica &&
+			IsLocalReferenceTableJoinPlan(plannedStmt))
+		{
+			if (IsMultiStatementTransaction())
```
Comment (marcocitus) [inline src/backend/distributed/executor/multi_executor.c]: maybe merge into above

File: src/backend/distributed/planner/distributed_planner.c [hunk 6]
```diff
@@ -118,7 +122,25 @@ distributed_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
...
+		 * table replicas.
+		 */
+		PrimaryNodeForGroup(GetLocalGroupId(), &hasReferenceTableReplica);
+
+		if (IsCoordinator() &&
+			hasReferenceTableReplica &&
+			IsLocalReferenceTableJoin(parse, rangeTableList))
+		{
+			needsDistributedPlanning = false;
+			UpdateReferenceTablesWithShard((Node *) parse, NULL);
```
Comment (marcocitus) [inline src/backend/distributed/planner/distributed_planner.c]: would appreciate a comment here

File: src/backend/distributed/planner/distributed_planner.c [hunk 7]
```diff
@@ -118,7 +122,25 @@ distributed_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
...
+
+		/*
+		 * All groups that have pg_dist_node entries, also have reference
+		 * table replicas.
+		 */
+		PrimaryNodeForGroup(GetLocalGroupId(), &hasReferenceTableReplica);
+
+		if (IsCoordinator() &&
+			hasReferenceTableReplica &&
+			IsLocalReferenceTableJoin(parse, rangeTableList))
```
Comment (marcocitus) [inline src/backend/distributed/planner/distributed_planner.c]: Maybe we can put these checks in a function?

Comment (pykello) [reply src/backend/distributed/planner/distributed_planner.c]: done

File: src/backend/distributed/planner/distributed_planner.c (hunk 6 above)
Comment (pykello) [reply src/backend/distributed/planner/distributed_planner.c]: done

File: src/backend/distributed/executor/multi_executor.c (hunk 5 above)
Comment (pykello) [reply src/backend/distributed/executor/multi_executor.c]: done

File: src/backend/distributed/planner/distributed_planner.c [hunk 8]
```diff
@@ -1770,3 +1787,172 @@ HasUnresolvedExternParamsWalker(Node *expression, ParamListInfo boundParams)
...
+	return hasLocalTable && hasReferenceTable;
+}
+
+
+/*
+ * QueryIsNotSimpleSelect returns true if node is a query which modifies or
+ * marks for modifications.
+ */
+static bool
+QueryIsNotSimpleSelect(Node *node)
```
Comment (onderkalaci) [inline src/backend/distributed/planner/distributed_planner.c]: what about modifying CTEs?

Comment (pykello) [reply src/backend/distributed/planner/distributed_planner.c]: Modifying CTEs contain queries which are not-select. In `IsLocalReferenceTableJoin` we recursively search for those by using `FindNodeCheck`.

**Pull request merged**